frontend/src/metabase/embedding-sdk/theme/dynamic-css-vars.ts

```diff
@@ -0,0 +1,86 @@
+import { css } from "@emotion/react";
+
+import { alpha, darken, isDark, isLight, lighten } from "metabase/lib/colors";
+import type { MantineTheme } from "metabase/ui";
+
+import type {
+  ColorOperation,
+  SourceColorKey,
+} from "../types/private/css-variables";
+
+import { DYNAMIC_CSS_VARIABLES } from "./dynamic-css-vars-config";
+
+const isColorDefined = (color: string) =>
+  color && color !== "transparent" && color !== "unset";
+
+/**
+ * Determine if the current color scheme is dark based on the palette.
+ */
+export function getIsDarkThemeFromPalette(theme: MantineTheme) {
+  const backgroundColor = theme.fn.themeColor("background");
+  const foregroundColor = theme.fn.themeColor("text-dark");
+
+  // Dark foreground color indicates a dark theme.
+  if (isColorDefined(backgroundColor)) {
+    return isDark(backgroundColor);
+  }
+
+  // Light foreground color indicates a dark theme.
+  if (isColorDefined(foregroundColor)) {
+    return isLight(foregroundColor);
+  }
+
+  return false;
+}
+
+/**
+ * Dynamically-generated CSS variables based on the theme.
+ * These colors are derived from the palette, with a configured tint and shade percentage.
+ */
+export function getDynamicCssVariables(theme: MantineTheme) {
+  const isDarkTheme = getIsDarkThemeFromPalette(theme);
+
+  const mappings = Object.entries(DYNAMIC_CSS_VARIABLES)
+    .map(([cssVar, config]) => {
+      let colorKey: SourceColorKey | null = null;
+      let operation: ColorOperation | null = null;
+
+      if (typeof config.source === "string") {
+        colorKey = config.source;
+      } else if (isDarkTheme && config.source.dark) {
+        colorKey = config.source.dark;
+      } else if (!isDarkTheme && config.source.light) {
```

Comment:
looking at the config, we don't have this one, I'm not sure what's the intent for having these conditions.

Can you explain more why we need these? They look overly complicated for me.

Another way to solve this problem for me is to design a new config type.

we can always force it to be like
```ts
type DynamicColorConfig = {
  dark: {
    source: string
    lighten: number
  },
  light: {
    source: string
    darken: number
  }
}
```

You will repeat `source` twice for colors that has the same source, but we won't have special logic that `source` are different types.

Reply:
@WiNloSt Great suggestion, repeating `source` twice for light and dark is the way go to here. The need for source-dependent color is because we need to sample the color differently for light and dark mode in some cases (e.g. brand for light mode, background for dark mode). I simplified the logic.

Reply:
I simplified the logic in https://github.com/metabase/metabase/pull/52513/commits/65a0c319c3e38744626c2f7c1a2bbaec5d3061d4